assets/src/state/pagePanelState.ts

```diff
@@ -0,0 +1,218 @@
+import { Vehicle, Ghost } from "../realtime"
+import { Action } from "../state"
+
+/**
+ * Valid "slots" in {@link ViewState}
+ */
+export enum PagePath {
+  Ladders = "/",
+  Shuttles = "/shuttle-map",
+  Search = "/search",
+  SearchMaps = "/map",
+}
+
+export function isPagePath(path: string): path is PagePath {
+  return Object.values(PagePath).includes(path as PagePath)
+}
+
+/**
+ * Keeps track of the current slot in use, and stores the state for each slot.
+ */
+export interface ViewState {
+  currentPath: PagePath
+  state: Record<PagePath, PageViewState>
+}
+
+/**
+ * Valid {@link PageViewState.openView} states
+ */
+export enum OpenView {
+  None = 1,
+  Swings,
+  Late,
+  NotificationDrawer,
+}
+
+/**
+ * State saved in each {@link ViewState} "slot".
+ */
+export type PageViewState = {
+  openView: OpenView
+  previousView: OpenView
+  selectedVehicleOrGhost: VehicleType
+}
+
+export const initialPageState = {
+  openView: OpenView.None,
+  previousView: OpenView.None,
+  selectedVehicleOrGhost: undefined,
+}
+
+export const initialPageViewState: ViewState = {
+  currentPath: PagePath.Ladders,
+  state: {
+    [PagePath.Ladders]: initialPageState,
+    [PagePath.Shuttles]: initialPageState,
+    [PagePath.Search]: initialPageState,
+    [PagePath.SearchMaps]: initialPageState,
+  },
+}
+
+export type VehicleType = Vehicle | Ghost | null | undefined
+
+//#region Reducers
+export const openViewReducer = (
+  currentState: ViewState,
+  action: Action
+): ViewState => {
+  let currentPath = currentState.currentPath
+  if (action.type === "SET_CURRENT_PATH") {
+    currentPath = action.path
+  }
+
+  return {
+    currentPath,
+    state: {
+      ...currentState.state,
+      [currentState.currentPath]: openViewPanelReducer(
+        currentState.state[currentState.currentPath],
+        action
+      ),
+    },
+  }
+}
+
+const openViewPanelReducer = (
+  state: PageViewState,
+  action: Action
+): PageViewState => {
+  const { openView, previousView, selectedVehicleOrGhost } = state
+  switch (action.type) {
+    case "SET_CURRENT_PATH": {
+      return {
+        openView: OpenView.None,
+        previousView: OpenView.None,
+        selectedVehicleOrGhost,
+      }
+    }
+    case "OPEN_NOTIFICATION_DRAWER":
+      return openView === OpenView.NotificationDrawer
+        ? state
```

Comment:
**praise:** I like the refactor in eb4969e, it definitely makes this code a bit more readable.